src/audio/AudioEngine.h

```diff
@@ -1,66 +1,177 @@
 
-    /** The AudioEngine represents the target OS sound system.
-     *
-     * The engine class matches the OpenAL device level.
+    enum class Format {
+        Mono,
+        Stereo
+    };
+
+    /**
+     * @brief Sounds represent audio sources.
+     * 
+     * All of the information about a sound should be re-set before it is played
+     * after it had been stopped.
+     * 
+     */
+    class Sound {
+    public:
+        virtual ~Sound() = default;
+
+        /**
+         * @brief Set sound pitch, must be positive
+         * 
+         * 1 means original pitch, values < 1 mean lower, > 1 mean higher.
+         * 
+         * @return float 
+         */
+        virtual void pitch(float) = 0;
+
+        /**
+         * @brief Set sound gain
+         * 
+         * 1 means original volume, <1 means lower gain, >1 means higher.
+         * 
+         * @return float 
+         */
+        virtual void gain(float) = 0;
+
+        /**
+         * @brief Set the maximum distance from which this sound can be heard.
+         *        Must be positive.
+         * 
+         */
+        virtual void maxDistance(float) = 0;
+        
+        /**
+         * @brief Set the coordinates of the sound's position.
+         * 
+         */
+        virtual void position(const Math::float3&) = 0;
+
+        /**
+         * @brief Set the components of the sound's velocity.
+         * 
+         */
+        virtual void velocity(const Math::float3&) = 0;
+
+        /**
+         * @brief Set the componenets of the sound's direction.
+         * 
+         */
+        virtual void direction(const Math::float3&) = 0;
+
+        /**
+         * @brief Set whether the sound's position is relative to the listener.
+         * 
+         */
+        virtual void relative(bool) = 0;
+
+        /**
+         * @brief Set wheter the sound should loop back to the beginning once it
+         *        finishes playing.
+         * 
+         * This is ignored in the case of streming sounds
+         */
+        virtual void looping(bool) = 0;
+
+        /**
+         * @brief Get the playback state of this sound.
+         * 
+         * @return State 
+         */
+        virtual State state() = 0;
+        
+        /**
+         * @brief Starts sound playback.
+         * 
+         * This method is idempotent: if the sound was already playing,
+         * nothing happens.
+         */
+        virtual void play() = 0;
+        /**
+         * @brief Pauses sound playback.
+         * 
+         * This method is idempotent: if the sound was already paused,
+         * nothing happens.
+         */
+        virtual void pause() = 0;
+        /**
+         * @brief Stops sound playback.
+         * 
+         * Playback will resume from the beginning.
+         * This method is idempotent: if the sound was already stopped,
+         * nothing happens.
+         */
+        virtual void stop() = 0;
+    };
+
+    using SoundPtr = std::shared_ptr<Sound>;
```

Comment:
nit: it should be doable with unique_ptr

Reply:
I wasn't able to do it with a `unique_ptr` because OpenGLAudioEngine keeps reference of the sounds that have been created

Reply:
Is the destruction order of `Sound` and `OpenGLAudioEngine` unspecified? Can `Sound` live without `OpenGLAudioEngine`?

Reply:
Considering that `OpenGLAudioEngine` is basically guaranteed to never be destroyed (it should be destroyed when the application closes), `Sound`s are always going to die before the engine. The issue is that the `Engine` gives out `Sound`s to its users, and these objects must stay alive even if they go out of scope, otherwise the sounds would stop playing. So the `Engine` keeps reference of all created sounds and destroys them when they are not tracked anymore by anyone else and they are not playing